source/extensions/filters/common/expr/library/custom_library.cc

```diff
@@ -0,0 +1,109 @@
+#include "source/extensions/filters/common/expr/library/custom_library.h"
+
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/extensions/rbac/custom_library_config/v3/custom_library.pb.h"
+#include "envoy/extensions/rbac/custom_library_config/v3/custom_library.pb.validate.h"
+#include "envoy/registry/registry.h"
+
+#include "source/extensions/filters/common/expr/library/custom_vocabulary.h"
+
+#include "eval/public/cel_function_adapter.h"
+#include "eval/public/cel_value.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Expr {
+namespace Library {
+
+void ThrowException(absl::string_view function_name, absl::Status status);
+
+void CustomLibrary::FillActivation(Activation* activation, Protobuf::Arena& arena,
+                                   const StreamInfo::StreamInfo& info,
+                                   const Http::RequestHeaderMap* request_headers,
+                                   const Http::ResponseHeaderMap* response_headers,
+                                   const Http::ResponseTrailerMap* response_trailers) {
+  request_headers_ = request_headers;
+  response_headers_ = response_headers;
+  response_trailers_ = response_trailers;
+  // words
+  activation->InsertValueProducer(
+      CustomVocabularyName, std::make_unique<CustomVocabularyWrapper>(
+                                arena, info, request_headers, response_headers, response_trailers));
+  // functions
+  absl::Status status;
```

Comment:
Perhaps explicitly call this out as lazy functions, and mention that the real benefit here is that they can be built be constructed at evaluation time, and so can be, e.g. influenced by the input params (headers etc).